tfx/components/example_validator/executor_test.py

```diff
@@ -19,55 +19,150 @@
+        os.path.join(self.validation_output.uri, 'anomalies.pbtxt'),
         anomalies_pb2.Anomalies())
     self.assertNotEqual(0, len(anomalies.anomaly_info))
+    return anomalies
+
+  def testDo(self):
+    input_dict = {
+        executor.STATISTICS_KEY: [self.eval_stats_artifact],
+        executor.SCHEMA_KEY: [self.schema_artifact],
+    }
+    output_dict = {
+        executor.ANOMALIES_KEY: [self.validation_output],
+    }
+
+    exec_properties = {}
+
+    self._test_do(input_dict, output_dict, exec_properties)
     # TODO(zhitaoli): Add comparison to expected anomolies.
 
+  def testDoSkewDetection(self):
+    # Read in the Chicago Taxi data.csv and intentionally skew it to keep only Cash trips
+    original_csv_path = os.path.join(self.source_data_dir,
+                                     'external/csv/data.csv')
+    skewed_data = pd.read_csv(original_csv_path)
+    skewed_data = skewed_data.loc[skewed_data['payment_type'] == 'Cash']
+
+    # Write the skewed df to .csv
+    tmp_dir = tempfile.mkdtemp()
+    skewed_dir = os.path.join(tmp_dir, 'skewed')
+    os.mkdir(skewed_dir)
+    skewed_path = os.path.join(skewed_dir, 'skewed_data.csv')
+    skewed_data.to_csv(skewed_path)
+
+    # TFX pipeline components to generate a schema and tfrecords file for the skewed data
+    examples = external_input(skewed_dir)
+    example_gen = CsvExampleGen(input=examples)
+    statistics_gen = components.StatisticsGen(
+        examples=example_gen.outputs['examples'])
+    schema_gen = components.SchemaGen(
+        statistics=statistics_gen.outputs['statistics'],
+        infer_feature_shape=False)
+
+    p = pipeline.Pipeline(
+        pipeline_name='skewed_chicago_taxi_beam',
+        pipeline_root=tmp_dir,
+        components=[example_gen, statistics_gen, schema_gen],
+        metadata_connection_config=metadata.sqlite_metadata_connection_config(
+            os.path.join(tmp_dir, 'metadata.db')))
+
+    BeamDagRunner().run(p)
+
+    # Construct the skewed statsitics artifact from pipeline outputs
+    component_dir = os.path.join(tmp_dir, 'StatisticsGen/statistics')
+    component_num = os.listdir(component_dir)[0]
+    skewed_statistics = standard_artifacts.ExampleStatistics()
+    skewed_statistics.uri = os.path.join(component_dir, component_num, 'eval')
+
+    # Set a skew comparator for the 'payment_type' feature
+    tmp_schema_pbtxt_dir = os.path.join(tmp_dir, 'SkewedSchemaGen')
+    os.mkdir(tmp_schema_pbtxt_dir)
+    tmp_schema_pbtxt_path = os.path.join(tmp_schema_pbtxt_dir, 'schema.pbtxt')
+    original_schema_pbtxt_path = os.path.join(self.schema_artifact.uri,
+                                              'schema.pbtxt')
+    shutil.copy2(original_schema_pbtxt_path, tmp_schema_pbtxt_path)
+
+    f = open(tmp_schema_pbtxt_path, "r")
+    contents = f.readlines()
+    f.close()
+
+    skew_comparator_str = '  skew_comparator: { infinity_norm: { threshold: .1} }'
+    contents.insert(12, skew_comparator_str)
```

Comment:
This is brittle and could break if the line number changes... consider reading the file into proto, doing the modification in the proto, and then writing it back.

Reply:
Done. I read the file in proto, add the skew_comparator and then write it back.